### ISSUE TITLE
[ci] Bedrock is routinely failing in the CI system.

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#8533]
Since a few days, `ci-bedrock2` is routinely failing in the CI system, due to what it looks like a memory exhaustion problem. Example: https://gitlab.com/coq/coq/-/jobs/100472191

This should be mitigated or `ci-bedrock` will be put in the allow-failure category. Possible mitigation strategies include: removing the problematic files from the ci build, forcing `-j1`, removing parallelism from the Coq tactics, or using a runner tag to request runners with more memory.

cc: @JasonGross @samuelgruetter

**Comment by Zimmi48** [comment]
What has been done previously when fiat-crypto was hitting a similar wall was to split the target into two, run the first with `-j2` and the second with `-j1`. This did the trick.

**Comment by ejgallego** [comment]
Indeed, -j2 may be the culprit here but however some workers only have 2Gb of RAM so that's not a lot.

**Comment by samuelgruetter** [comment]
@ejgallego does `-j1` fix the problem?
(Another culprit might be the `par:` in [DecodeEncode.v](https://github.com/mit-plv/riscv-coq/blob/21993a3401647c7bc6da9b6f7bd0fa978d61001b/src/proofs/DecodeEncode.v#L267), but that one is critical to keep the development "round trip time" short)
And btw, can't you just increase the RAM limit?

**Comment by samuelgruetter** [comment]
> I wonder if it could be disabled just for the CI.

This would be bad:
*    It makes the claim "it builds locally if and only if it builds on CI" less true
*    It complicates the Makefile
*    It means we miss an opportunity to have a bigger-than-toy-example regression test for Coq's `par:` feature

> How can we do that? As far as I can tell runners come with a defined amount of RAM, which is usually enough for the whole test-suite [which is not precisely a toy these days]

According to the [CircleCI docs](https://circleci.com/docs/2.0/configuration-reference/#resource_class), this is possible.

I want to use Coq as a tool to do research on what will be possible in software verification in a few years from now. I hope that by then, Coq's memory usage will be lower, and RAM will be even cheaper than now. Therefore, we should really not be held back by low RAM limits today!

**Comment by ejgallego** [comment]
@samuelgruetter a few points:
- first, we are not using CircleCI, but GitLab free runners, these come with those limitations,
- second, we can agree on your views on the future of software verification, but we have to be more pragmatic here. As of today, Coq's CI has the resources it has, we welcome improvements / donations, but in the short term, it means that we are able to fit your development into our current infrastructure or we will sadly have to drop it from the CI suite.
- once we assume this problem has to be fixed, I think from our part we are open to whatever works for you, so feel free to propose an alternative. Meanwhile, we will move `ci-bedrock2` to the `allow-fail` category as it is causing too much noise.

**Comment by andres-erbsen** [comment]
The documentation [suggests](http://web.mit.edu/jgross/Public/tmp/doc/sphinx/_build/html/addendum/parallel-proof-processing.html#limiting-the-number-of-parallel-workers) that `coqworkmgr` can be used to limit the number of parallel jobs. Would that be applicable here? Alternatively, `find . -name '*.v' | xargs sed -i 's/par:/all:/'` might be the appropriate level of kludge to apply here.

**Comment by ejgallego** [comment]
@andres-erbsen , I am not sure how `coqworkmgr` is supposed to work (cc: @gares ) I was under the impression that you had to set the flag when calling `coqc` itself.

> Alternatively, `find . -name '*.v' | xargs sed -i 's/par:/all:/'` might be the appropriate level of kludge to apply here.

That looks like a fairly ugly klugde indeed, I'd prefer we find a cleaner solution, but if that's what works for you OK then.

**Comment by samuelgruetter** [comment]
Couldn't this be as [easy](https://docs.gitlab.com/runner/configuration/advanced-configuration.html#the-runners-docker-section) as adding something like

```
runners:
  docker:
    memory: "8G"
```

to coq's `.gitlab-ci.yml`?

**Comment by ejgallego** [comment]
> Couldn't this be as [easy](https://docs.gitlab.com/runner/configuration/advanced-configuration.html#the-runners-docker-section) as adding something like

I'm afraid not. What you are looking at is at the configuration of the runner daemon itself.

**Comment by Zimmi48** [comment]
@ejgallego Note that @coqbot itself needs to be adapted to not report about bedrock failures if the decision taken is to move it in `allow_fail`.

**Comment by ejgallego** [comment]
> @ejgallego Note that @coqbot itself needs to be adapted to not report about bedrock failures if the decision taken is to move it in `allow_fail`.

Then we may just disable it entirely until the fix is implemented, as it is really bugging us.

**Comment by samuelgruetter** [comment]
> Alternatively, `find . -name '*.v' | xargs sed -i 's/par:/all:/'` might be the appropriate level of kludge to apply here.

This is certainly ugly, but if this problem can be "solved" just by putting the above line somewhere in `.gitlab-ci.yml`, let's do that, as long as we don't have better solutions at hand.

**Comment by ejgallego** [comment]
That seems too radical, if we have to go this route I suggest we do that only for the problematic file.

**Comment by samuelgruetter** [comment]
> That seems too radical, if we have to go this route I suggest we do that only for the problematic file.

If you specify each file separately, you'll have to update `.gitlab-ci.yml` each time someone adds a new memory-intensive proof using `par:` to bedrock2. So far the whole codebase contains only 1 `par:`, so this might happen only rarely. Up to you to decide :wink:

**Comment by ejgallego** [comment]
Well, the thing is that if Bedrock is going to become a memory-intensive development [which of course it is a 100% legitimate thing to do] we may really need to do a toned-down version for CI until we can have a better infrastructure. I agree that 2GiB runners is a bit low but so far this is what we have.

It is your development folks, if you think the full `par:` replacement is the way to go let's go ahead.

But first I'd like to hear from @gares , is there are way to limit `par:` so it uses only one thread? I did not understand that section of the manual.

**Comment by gares** [comment]
You can pass `-async-proofs-tac-j $N` to cap what `par:` spwans, but IIRC the default is only 2.

`coqworkmgr` can be used to cap globally, you could have `par:` capped at 8, do a make -j 5 but globally cap workers to 20. I guess here we should -j1 and -tac-j 1.

I'd like to keep `par:` in place, even if capped at 1 worker.

**Comment by ejgallego** [comment]
Thanks @gares , so what do you recommend? Should we cap `par:` with `-async-proofs-tac-j 1` or do you think that `coqworkmgr` would work better.

Meanwhile, @samuelgruetter , I guess we could try adding an option to beckrock's makefile so we can set `-async-proofs-tac-j $PAR_JOBS` and we will set `PAR_JOBS` to 1 in the CI.

**Comment by ejgallego** [comment]
Indeed I'd rather keep the `par:` in place.

**Comment by gares** [comment]
I'd pass `-async-proofs-tac-j 1` and then try `make -j2` or `make -j1`.

If it uses coq_makefile, then setting `COQEXTRAFLAGS` should suffice

**Comment by samuelgruetter** [comment]
The subproject in question (riscv-coq) does use `coq_makefile`, so can you just add

```
COQEXTRAFLAGS+=-async-proofs-tac-j 1
```

somewhere in `gitlab-ci.yml` and that's it?

**Comment by gares** [comment]
`make -f Makefile.coq -j2 COQEXTRAFLAGS="-async-proofs-tac-j 1"` should do it

**Comment by Zimmi48** [comment]
Not in `.gitlab-ci.yml` for sure but rather in `dev/ci/ci-bedrock2.sh`.
But anyways, sorry for realizing this late but there is another solution since we have the pyrolyse runner: we could tag this job to run only on Pyrolyse although I need to understand how to run more than one build in parallel on a single runner. Pyrolyse has 32 GB of RAM and 32 CPUs. We could at least run 8 jobs in parallel.

**Comment by ejgallego** [comment]
> we could tag this job to run only on Pyrolyse although I need to understand how to run more than one build in parallel on a single runner

We had more capable runners in the past, see issue #6962

In the end, that's a bit of a pain for people willing to use their own gitlab; I feel that at least for this particular issue we should try to limit the amount of workers.

**Comment by ejgallego** [comment]
So I added a `coqworkmgr -j 1` call here: https://gitlab.com/ejgallego/coq/-/jobs/101252135 but however something seems veeeery wrong, a normal bedrock build takes around 23 minutes, this build is taking close to 80 minutes and counting...

**Comment by samuelgruetter** [comment]
The build https://gitlab.com/ejgallego/coq/-/jobs/101252135 built all files (`Fibonacci.v` and `EditDistExample.v` are the last files in the dependency graph), but after building all Coq files, it seems to hang on `Pulling docker image gitlab/gitlab-runner-helper:x86_64-cde4a2d1 ...`

**Comment by ejgallego** [comment]
It just timed out, so something is very wrong when we call `coqworkmgr`

**Comment by ejgallego** [comment]
It seems all runners now come with 4GiB of RAM so tentatively closing this. Bedrock upstream tho should be careful with the target they want to test in our CI.

Unfortunately we are not in shape to offer very powerful runners in the medium-term so they should keep CI testing under reasonable constraints.

**Comment by ejgallego** [comment]
Yeah, the sooner I speak...

**Comment by Zimmi48** [comment]
And another: https://gitlab.com/coq/coq/-/jobs/104420598

**Comment by Zimmi48** [comment]
It is still an option to have these tagged to run on pyrolyse. As for the allow_failure option, it is now supported by @coqbot.

**Comment by ejgallego** [comment]
> It is still an option to have these tagged to run on pyrolyse.

We can do that in a similar way to Windows builds, but I'd rather try not to introduce yet another special category.

**Comment by Zimmi48** [comment]
As you prefer...

**Comment by ejgallego** [comment]
My fear is that if we open the mega-resource door to CI we won't be able to ever close it.

**Comment by Zimmi48** [comment]
I get your fear but we have full control on what we decide can get special tagging, and, moreover, we can even say that bedrock2 on pyrolyse will be temporary (we'll remove it if pyrolyse gets only for Windows builds).

**Comment by ejgallego** [comment]
I see a problem with this, in the sense that it would imply the removal of `bedrock` later on then?

**Comment by Zimmi48** [comment]
Yep, unless in the meantime we have found a solution to tag builds needing more memory to use only some kind of free runners.

**Comment by samuelgruetter** [comment]
It seems that in Coq 8.8.1, the variable to use is not `COQEXTRAFLAGS`, but `OTHERFLAGS`.
In the riscv-coq directory, if I do

```
echo 'OTHERFLAGS=-async-proofs-tac-j 1' > Makefile.coq.all.local
make VERBOSE=1 -j2 all
```

it displays commands being run like the following, which looks good:

```
"coqc"   -q  -Q ../bbv/theories bbv -R src riscv -async-proofs-tac-j 1 src/ExecuteM64.v 
```

except that they fail with the following error:

```
coqc: -async-proofs-tac-j: no such file or directory
```

So I thought, maybe I have to put this argument first, so manually ran this:

```
"coqc" -async-proofs-tac-j 1  -q  -Q ../bbv/theories bbv -R src riscv src/ExecuteM.v
```

which gave the same error.

Two questions:

1. Why do I get this error?
2. Is there any change needed in the bedrock2 project, or can you just add something like `echo 'OTHERFLAGS=-async-proofs-tac-j 1' > Makefile.coq.all.local` into `dev/ci/ci-bedrock2.sh`?

**Comment by ejgallego** [comment]
Hi @samuelgruetter , I guess I'll invoke @gares here. Indeed, I suggest we put the line in `dev/ci/ci-bedrock2.sh` so this is only enabled for the CI build. Thanks!

**Comment by Zimmi48** [comment]
This stuff is really not documented well enough.

**Comment by ejgallego** [comment]
pinging @gares again; for now we are going to place bedrock2 in "allow_failure" mode; this means PRs are allowed to break it; hoping to see it back soon.

**Comment by gares** [comment]
I think coqc is stupid: it interprets all unknown, to it, flags as files...

**Comment by gares** [comment]
Replacing coqc with coqtop -compile may be a viable work around

**Comment by samuelgruetter** [comment]
Looks like this is a bug due to code duplication between [tools/coqc.ml](https://github.com/coq/coq/blob/master/tools/coqc.ml) and [toplevel/coqargs.ml](https://github.com/coq/coq/blob/master/toplevel/coqargs.ml)...

**Comment by ejgallego** [comment]
Well indeed `coqc` is very redudant, maybe it is time we make it a link to `coqtop`? WDYT @herbelin ?

**Comment by gares** [comment]
I think `coqc` should be `git rm` and `coqtop` should check if `executable_name` is `coqc` and in that case enable `-compile`.

**Comment by ejgallego** [comment]
> I think `coqc` should be `git rm` and `coqtop` should check if `executable_name` is `coqc` and in that case enable `-compile`. (and indeed provide a link/copy for backward compatibility)

That's a good idea, but we can go even further and really split `coqtop` on two, like `ocaml` `ocamlc` are, WDTY?

The compiler part is fully orthogonal from the toplevel part these days.

**Comment by herbelin** [comment]
> Well indeed coqc is very redundant, maybe it is time we make it a link to coqtop?

No opinion (but I suppose that many users would like their `coqc`-based scripts continuing to work).

**Comment by ejgallego** [comment]
> No opinion (but I suppose that many users would like their `coqc`-based scripts continuing to work).

Or course we would still install a binary/link. I was talking about the file `tools/coqc.ml`

**Comment by gares** [comment]
You can also split, but please share the code parsing the common arguments (that is 99% of them I'm afraid). In other words, I'm no so sure the refactoring is going to pay back.

**Comment by ejgallego** [comment]
> You can also split, but please share the code parsing the common arguments (that is 99% of them I'm afraid). In other words, I'm no so sure the refactoring is going to pay back.

Of course we would share the argument code, it is designed for that. The main advantage of the refactoring is that a few strange cases would be impossible now by design. Right now there is a significant mess in the code-flow due to having to support both cases in the same codepath.

**Comment by samuelgruetter** [comment]
Would you mind continuing the command line arguments discussion in a separate issue? :arrow_right:  https://github.com/coq/coq/issues/8683

I'd like to stay subscribed to the bedrock2 related discussion, without getting distracted by all the command line argument parsing discussion :wink:

**Comment by ejgallego** [comment]
Thanks for filling the issue @samuelgruetter, do you think the CI problem can be addressed without fixing it?

**Comment by ejgallego** [comment]
https://github.com/coq/coq/issues/8683 has been fixed, can we try to make progress on this and re-enable bedrock?

**Comment by ejgallego** [comment]
Thanks @samuelgruetter for the great summary, which looks pretty accurate to me, and sorry for the long time it took fixing the command line argument parsing.

Indeed hopefully we could add that to the `ci-bedrock` script so we get more reliable behavior on the CI.

**Comment by samuelgruetter** [comment]
Update: `-async-proofs-tac-j` still doesn't work, see https://github.com/coq/coq/issues/9484

**Comment by ejgallego** [comment]
Closing this thanks to #9534
